File: base/Sources/TabMaterials.ts

///if (is_paint || is_sculpt)

class TabMaterials {

	static draw = (htab: zui_handle_t) => {
		let mini: bool = config_raw.layout[layout_size_t.SIDEBAR_W] <= UIBase.sidebar_mini_w;
		mini ? TabMaterials.draw_mini(htab) : TabMaterials.draw_full(htab);
	}

	static draw_mini = (htab: zui_handle_t) => {
		zui_set_hovered_tab_name(tr("Materials"));

		zui_begin_sticky();
		zui_separator(5);

		TabMaterials.button_nodes();
		TabMaterials.button_new("+");

		zui_end_sticky();
		zui_separator(3, false);
		TabMaterials.draw_slots(true);
	}

	static draw_full = (htab: zui_handle_t) => {
		if (zui_tab(htab, tr("Materials"))) {
			zui_begin_sticky();
			zui_row([1 / 4, 1 / 4, 1 / 4]);

			TabMaterials.button_new(tr("New"));
			if (zui_button(tr("Import"))) {
				project_import_material();
			}
			TabMaterials.button_nodes();

			zui_end_sticky();
			zui_separator(3, false);
			TabMaterials.draw_slots(false);
		}
	}

	static button_nodes = () => {
		let ui: zui_t = UIBase.ui;
		if (zui_button(tr("Nodes"))) {
			UIBase.show_material_nodes();
		}
		else if (ui.is_hovered) zui_tooltip(tr("Show Node Editor") + ` (${config_keymap.toggle_node_editor})`);
	}

	static draw_slots = (mini: bool) => {
		let ui: zui_t = UIBase.ui;
		let slotw: i32 = Math.floor(51 * zui_SCALE(ui));
		let num: i32 = Math.floor(config_raw.layout[layout_size_t.SIDEBAR_W] / slotw);

		for (let row: i32 = 0; row < Math.floor(Math.ceil(project_materials.length / num)); ++row) {
			let mult: i32 = config_raw.show_asset_names ? 2 : 1;
			let ar: f32[] = [];
			for (let i = 0; i < num * mult; ++i) ar.push(1 / num);
			zui_row(ar);

			ui._x += 2;
			let off: f32 = config_raw.show_asset_names ? zui_ELEMENT_OFFSET(ui) * 10.0 : 6;
			if (row > 0) ui._y += off;

			for (let j: i32 = 0; j < num; ++j) {
				let imgw: i32 = Math.floor(50 * zui_SCALE(ui));
				let i: i32 = j + row * num;
				if (i >= project_materials.length) {
					zui_end_element(imgw);
					if (config_raw.show_asset_names) zui_end_element(0);
					continue;
				}
				let img: image_t = zui_SCALE(ui) > 1 ? project_materials[i].image : project_materials[i].image_icon;
				let imgFull: image_t = project_materials[i].image;

				// Highligh selected
				if (context_raw.material == project_materials[i]) {
					if (mini) {
						let w: f32 = ui._w / zui_SCALE(ui);
						zui_rect(0, -2, w - 2, w - 4, ui.t.HIGHLIGHT_COL, 3);
					}
					else {
						let off: i32 = row % 2 == 1 ? 1 : 0;
						let w: i32 = 50;
						if (config_raw.window_scale > 1) w += Math.floor(config_raw.window_scale * 2);
						zui_fill(-1,         -2, w + 3,       2, ui.t.HIGHLIGHT_COL);
						zui_fill(-1,    w - off, w + 3, 2 + off, ui.t.HIGHLIGHT_COL);
						zui_fill(-1,         -2,     2,   w + 3, ui.t.HIGHLIGHT_COL);
						zui_fill(w + 1,      -2,     2,   w + 4, ui.t.HIGHLIGHT_COL);
					}
				}

				///if krom_opengl
				ui.image_invert_y = project_materials[i].preview_ready;
				///end

				// Draw material icon
				let uix: f32 = ui._x;
				let uiy: f32 = ui._y;
				let tile: i32 = zui_SCALE(ui) > 1 ? 100 : 50;
				let imgh: f32 = mini ? UIBase.default_sidebar_mini_w * 0.85 * zui_SCALE(ui) : -1.0;
				let state = project_materials[i].preview_ready ?
					zui_image(img, 0xffffffff, imgh) :
					zui_image(resource_get("icons.k"), 0xffffffff, -1.0, tile, tile, tile, tile);

				// Draw material numbers when selecting a material via keyboard shortcut
				let is_typing: bool = ui.is_typing || UIView2D.ui.is_typing || UINodes.ui.is_typing;
				if (!is_typing) {
					if (i < 9 && operator_shortcut(config_keymap.select_material, shortcut_type_t.DOWN)) {
						let number: string = String(i + 1);
						let width: i32 = g2_font_width(ui.font, ui.font_size, number) + 10;
						let height: i32 = g2_font_height(ui.font, ui.font_size);
						g2_set_color(ui.t.TEXT_COL);
						g2_fill_rect(uix, uiy, width, height);
						g2_set_color(ui.t.ACCENT_COL);
						g2_draw_string(number, uix + 5, uiy);
					}
				}

				// Select material
				if (state == zui_state_t.STARTED && ui.input_y > ui._window_y) {
					if (context_raw.material != project_materials[i]) {
						context_select_material(i);
						///if is_paint
						if (context_raw.tool == workspace_tool_t.MATERIAL) {
							let _init = () => {
								base_update_fill_layers();
							}
							app_notify_on_init(_init);
						}
						///end
					}
					base_drag_off_x = -(mouse_x - uix - ui._window_x - 3);
					base_drag_off_y = -(mouse_y - uiy - ui._window_y + 1);
					base_drag_material = context_raw.material;
					// Double click to show nodes
					if (time_time() - context_raw.select_time < 0.25) {
						UIBase.show_material_nodes();
						base_drag_material = null;
						base_is_dragging = false;
					}
					context_raw.select_time = time_time();
				}

				// Context menu
				if (ui.is_hovered && ui.input_released_r) {
					context_select_material(i);
					let add: i32 = project_materials.length > 1 ? 1 : 0;

					UIMenu.draw((ui: zui_t) => {
						let m: SlotMaterialRaw = project_materials[i];

						if (UIMenu.menu_button(ui, tr("To Fill Layer"))) {
							context_select_material(i);
							base_create_fill_layer();
						}

						if (UIMenu.menu_button(ui, tr("Export"))) {
							context_select_material(i);
							BoxExport.show_material();
						}

						///if is_paint
						if (UIMenu.menu_button(ui, tr("Bake"))) {
							context_select_material(i);
							BoxExport.show_bake_material();
						}
						///end

						if (UIMenu.menu_button(ui, tr("Duplicate"))) {
							let _init = () => {
								context_raw.material = SlotMaterial.create(project_materials[0].data);
								project_materials.push(context_raw.material);
								let cloned: zui_node_canvas_t = json_parse(json_stringify(project_materials[i].canvas));
								context_raw.material.canvas = cloned;
								TabMaterials.update_material();
								history_duplicate_material();
							}
							app_notify_on_init(_init);
						}

						if (project_materials.length > 1 && UIMenu.menu_button(ui, tr("Delete"), "delete")) {
							TabMaterials.delete_material(m);
						}

						let base_handle: zui_handle_t = zui_nest(zui_handle("tabmaterials_0"), m.id, {selected: m.paint_base});
						let opac_handle: zui_handle_t = zui_nest(zui_handle("tabmaterials_1"), m.id, {selected: m.paint_opac});
						let nor_handle: zui_handle_t = zui_nest(zui_handle("tabmaterials_2"), m.id, {selected: m.paint_nor});
						let occ_handle: zui_handle_t = zui_nest(zui_handle("tabmaterials_3"), m.id, {selected: m.paint_occ});
						let rough_handle: zui_handle_t = zui_nest(zui_handle("tabmaterials_4"), m.id, {selected: m.paint_rough});
						let met_handle: zui_handle_t = zui_nest(zui_handle("tabmaterials_5"), m.id, {selected: m.paint_met});
						let height_handle: zui_handle_t = zui_nest(zui_handle("tabmaterials_6"), m.id, {selected: m.paint_height});
						let emis_handle: zui_handle_t = zui_nest(zui_handle("tabmaterials_7"), m.id, {selected: m.paint_emis});
						let subs_handle: zui_handle_t = zui_nest(zui_handle("tabmaterials_8"), m.id, {selected: m.paint_subs});
						UIMenu.menu_fill(ui);
						m.paint_base = zui_check(base_handle, tr("Base Color"));
						UIMenu.menu_fill(ui);
						m.paint_opac = zui_check(opac_handle, tr("Opacity"));
						UIMenu.menu_fill(ui);
						m.paint_nor = zui_check(nor_handle, tr("Normal"));
						UIMenu.menu_fill(ui);
						m.paint_occ = zui_check(occ_handle, tr("Occlusion"));
						UIMenu.menu_fill(ui);
						m.paint_rough = zui_check(rough_handle, tr("Roughness"));
						UIMenu.menu_fill(ui);
						m.paint_met = zui_check(met_handle, tr("Metallic"));
						UIMenu.menu_fill(ui);
						m.paint_height = zui_check(height_handle, tr("Height"));
						UIMenu.menu_fill(ui);
						m.paint_emis = zui_check(emis_handle, tr("Emission"));
						UIMenu.menu_fill(ui);
						m.paint_subs = zui_check(subs_handle, tr("Subsurface"));
						if (base_handle.changed ||
							opac_handle.changed ||
							nor_handle.changed ||
							occ_handle.changed ||
							rough_handle.changed ||
							met_handle.changed ||
							height_handle.changed ||
							emis_handle.changed ||
							subs_handle.changed) {
							MakeMaterial.parse_paint_material();
							UIMenu.keep_open = true;
						}
					}, 13 + add);
				}
				if (ui.is_hovered) {
					zui_tooltip_image(imgFull);
					if (i < 9) zui_tooltip(project_materials[i].canvas.name + " - (" + config_keymap.select_material + " " + (i + 1) + ")");
					else zui_tooltip(project_materials[i].canvas.name);
				}

				if (config_raw.show_asset_names) {
					ui._x = uix;
					ui._y += slotw * 0.9;
					zui_text(project_materials[i].canvas.name, zui_align_t.CENTER);
					if (ui.is_hovered) {
						if (i < 9) zui_tooltip(project_materials[i].canvas.name + " - (" + config_keymap.select_material + " " + (i + 1) + ")");
						else zui_tooltip(project_materials[i].canvas.name);
					}
					ui._y -= slotw * 0.9;
					if (i == project_materials.length - 1) {
						ui._y += j == num - 1 ? imgw : imgw + zui_ELEMENT_H(ui) + zui_ELEMENT_OFFSET(ui);
					}
				}
			}

			ui._y += mini ? 0 : 6;

			///if krom_opengl
			ui.image_invert_y = false; // Material preview
			///end
		}

		let in_focus: bool = ui.input_x > ui._window_x && ui.input_x < ui._window_x + ui._window_w &&
					    	 ui.input_y > ui._window_y && ui.input_y < ui._window_y + ui._window_h;
		if (in_focus && ui.is_delete_down && project_materials.length > 1) {
			ui.is_delete_down = false;
			TabMaterials.delete_material(context_raw.material);
		}
	}

	static button_new = (text: string) => {
		if (zui_button(text)) {
			let current: image_t = _g2_current;
			g2_end();
			context_raw.material = SlotMaterial.create(project_materials[0].data);
			project_materials.push(context_raw.material);
			TabMaterials.update_material();
			g2_begin(current);
			history_new_material();
		}
	}

	static update_material = () => {
		UIHeader.header_handle.redraws = 2;
		UINodes.hwnd.redraws = 2;
		UINodes.group_stack = [];
		MakeMaterial.parse_paint_material();
		UtilRender.make_material_preview();
		let decal: bool = context_raw.tool == workspace_tool_t.DECAL || context_raw.tool == workspace_tool_t.TEXT;
		if (decal) UtilRender.make_decal_preview();
	}

	static update_material_pointers = (nodes: zui_node_t[], i: i32) => {
		for (let n of nodes) {
			if (n.type == "MATERIAL") {
				if (n.buttons[0].default_value == i) {
					n.buttons[0].default_value = 9999; // Material deleted
				}
				else if (n.buttons[0].default_value > i) {
					n.buttons[0].default_value--; // Offset by deleted material
				}
			}
		}
	}

	static accept_swatch_drag = (swatch: swatch_color_t) => {
		context_raw.material = SlotMaterial.create(project_materials[0].data);
		for (let node of context_raw.material.canvas.nodes) {
			if (node.type == "RGB" ) {
				node.outputs[0].default_value = [
					color_get_rb(swatch.base) / 255,
					color_get_gb(swatch.base) / 255,
					color_get_bb(swatch.base) / 255,
					color_get_ab(swatch.base) / 255
				];
			}
			else if (node.type == "OUTPUT_MATERIAL_PBR") {
				node.inputs[1].default_value = swatch.opacity;
				node.inputs[2].default_value = swatch.occlusion;
				node.inputs[3].default_value = swatch.roughness;
				node.inputs[4].default_value = swatch.metallic;
				node.inputs[7].default_value = swatch.height;
			}
		}
		project_materials.push(context_raw.material);
		TabMaterials.update_material();
		history_new_material();
	}

	static delete_material = (m: SlotMaterialRaw) => {
		let i: i32 = project_materials.indexOf(m);
		for (let l of project_layers) if (l.fill_layer == m) l.fill_layer = null;
		history_delete_material();
		context_select_material(i == project_materials.length - 1 ? i - 1 : i + 1);
		project_materials.splice(i, 1);
		UIBase.hwnds[1].redraws = 2;
		for (let m of project_materials) TabMaterials.update_material_pointers(m.canvas.nodes, i);
	}
}

///end

File: paint/sources/tab_materials.c

#include "global.h"

i32 _tab_materials_draw_slots;
i32 tab_materials_drag_pos = -1;

void tab_materials_button_nodes() {
	if (ui_button(tr("Nodes"), UI_ALIGN_CENTER, "")) {
		ui_base_show_material_nodes();
	}
	else if (ui->is_hovered) {
		ui_tooltip(string("%s (%s)", tr("Show Node Editor"), (char *)any_map_get(config_keymap, "toggle_node_editor")));
	}
}

void tab_materials_draw_slots_update_fill_layers(void *_) {
	layers_update_fill_layers();
}

void tab_materials_update_material() {
	ui_header_handle->redraws = 2;
	ui_nodes_hwnd->redraws    = 2;
	gc_unroot(ui_nodes_group_stack);
	ui_nodes_group_stack = any_array_create_from_raw((void *[]){}, 0);
	gc_root(ui_nodes_group_stack);
	make_material_parse_paint_material(true);
	util_render_make_material_preview();
	bool decal = context_is_decal();
	if (decal) {
		util_render_make_decal_preview();
	}
	base_update_workflow();
}

void tab_materials_draw_slots_duplicate(void *_) {
	i32 i               = _tab_materials_draw_slots;
	g_context->material = slot_material_create(project_materials->buffer[0]->data, NULL);
	any_array_push(project_materials, g_context->material);
	ui_node_canvas_t *cloned    = util_clone_canvas(project_materials->buffer[i]->canvas);
	g_context->material->canvas = cloned;
	tab_materials_update_material();
	history_duplicate_material();
}

void tab_materials_update_material_pointers(ui_node_t_array_t *nodes, i32 i) {
	for (i32 i = 0; i < nodes->length; ++i) {
		ui_node_t *n = nodes->buffer[i];
		if (string_equals(n->type, "MATERIAL")) {
			if (n->buttons->buffer[0]->default_value->buffer[0] == i) {
				n->buttons->buffer[0]->default_value->buffer[0] = 9999; // Material deleted
			}
			else if (n->buttons->buffer[0]->default_value->buffer[0] > i) {
				n->buttons->buffer[0]->default_value->buffer[0]--; // Offset by deleted material
			}
		}
	}
}

void tab_materials_delete_material(slot_material_t *m) {
	i32 i = array_index_of(project_materials, m);
	for (i32 i = 0; i < project_layers->length; ++i) {
		slot_layer_t *l = project_layers->buffer[i];
		if (l->fill_layer == m) {
			l->fill_layer = NULL;
		}
	}
	history_delete_material();
	context_select_material(i == project_materials->length - 1 ? i - 1 : i + 1);
	array_splice(project_materials, i, 1);
	ui_base_hwnds->buffer[1]->redraws = 2;
	for (i32 i = 0; i < project_materials->length; ++i) {
		slot_material_t *m = project_materials->buffer[i];
		tab_materials_update_material_pointers(m->canvas->nodes, i);
	}
	for (i32 i = 0; i < m->canvas->nodes->length; ++i) {
		ui_node_t *n = m->canvas->nodes->buffer[i];
		ui_viewnodes_on_node_remove(n);
	}
}

void tab_materials_draw_slots_menu() {
	i32              i = _tab_materials_draw_slots;
	slot_material_t *m = project_materials->buffer[i];

	if (ui_menu_button(tr("To Fill Layer"), "", ICON_SPHERE)) {
		context_select_material(i);
		layers_create_fill_layer(UV_TYPE_UVMAP, mat4_nan(), -1);
	}

	if (ui_menu_button(tr("Export"), "", ICON_EXPORT)) {
		context_select_material(i);
		box_export_show_material();
	}

	if (ui_menu_button(tr("Bake"), "", ICON_BAKE)) {
		context_select_material(i);
		box_export_show_bake_material();
	}

	if (ui_menu_button(tr("Duplicate"), "ctrl+d", ICON_DUPLICATE)) {
		sys_notify_on_next_frame(&tab_materials_draw_slots_duplicate, NULL);
	}

	if (project_materials->length > 1 && ui_menu_button(tr("Delete"), "delete", ICON_DELETE)) {
		tab_materials_delete_material(m);
	}

	if (g_config->experimental && ui_menu_button(tr("Inspect as Script"), "", ICON_SEARCH)) {
		node_shader_dump_to_script = true;
		context_select_material(i);
	}

	ui_handle_t *base_handle = ui_nest(ui_handle(__ID__), m->id);
	if (base_handle->init) {
		base_handle->b = m->paint_base;
	}

	ui_handle_t *opac_handle = ui_nest(ui_handle(__ID__), m->id);
	if (opac_handle->init) {
		opac_handle->b = m->paint_opac;
	}

	ui_handle_t *nor_handle = ui_nest(ui_handle(__ID__), m->id);
	if (nor_handle->init) {
		nor_handle->b = m->paint_nor;
	}

	ui_handle_t *occ_handle = ui_nest(ui_handle(__ID__), m->id);
	if (occ_handle->init) {
		occ_handle->b = m->paint_occ;
	}

	ui_handle_t *rough_handle = ui_nest(ui_handle(__ID__), m->id);
	if (rough_handle->init) {
		rough_handle->b = m->paint_rough;
	}

	ui_handle_t *met_handle = ui_nest(ui_handle(__ID__), m->id);
	if (met_handle->init) {
		met_handle->b = m->paint_met;
	}

	ui_handle_t *height_handle = ui_nest(ui_handle(__ID__), m->id);
	if (height_handle->init) {
		height_handle->b = m->paint_height;
	}

	ui_handle_t *emis_handle = ui_nest(ui_handle(__ID__), m->id);
	if (emis_handle->init) {
		emis_handle->b = m->paint_emis;
	}

	ui_handle_t *subs_handle = ui_nest(ui_handle(__ID__), m->id);
	if (subs_handle->init) {
		subs_handle->b = m->paint_subs;
	}

	ui_menu_separator();
	ui_menu_align();
	ui_menu_label(tr("Opacity Mode"), NULL);
	ui_menu_align();
	ui_handle_t *opac_mode_handle   = ui_handle(__ID__);
	opac_mode_handle->i             = m->paint_opac_mode;
	string_array_t *opac_mode_items = any_array_create_from_raw(
	    (void *[]){
	        tr("Alpha"),
	        tr("Translucency"),
	    },
	    2);
	m->paint_opac_mode = ui_inline_radio(opac_mode_handle, opac_mode_items, UI_ALIGN_LEFT);

	ui_menu_separator();
	ui_menu_align();
	ui_menu_label(tr("Channels"), NULL);
	ui_menu_align();
	ui_row2();
	m->paint_base = ui_check(base_handle, tr("Base Color"), "");
	m->paint_opac = ui_check(opac_handle, tr("Opacity"), "");

	if (g_config->workflow == WORKFLOW_PBR) {
		ui_row2();
		m->paint_nor    = ui_check(nor_handle, tr("Normal"), "");
		m->paint_height = ui_check(height_handle, tr("Height"), "");
		ui_row2();
		m->paint_rough = ui_check(rough_handle, tr("Roughness"), "");
		m->paint_met   = ui_check(met_handle, tr("Metallic"), "");
		ui_row2();
		m->paint_emis = ui_check(emis_handle, tr("Emission"), "");
		m->paint_subs = ui_check(subs_handle, tr("Subsurface"), "");
		m->paint_occ  = ui_check(occ_handle, tr("Occlusion"), "");
	}

	if (base_handle->changed || opac_handle->changed || nor_handle->changed || occ_handle->changed || rough_handle->changed || met_handle->changed ||
	    height_handle->changed || emis_handle->changed || subs_handle->changed || opac_mode_handle->changed) {
		make_material_parse_paint_material(true);
		if (opac_mode_handle->changed) {
			sys_notify_on_next_frame(util_render_make_material_preview, NULL);
		}
		ui_menu_keep_open = true;
	}
}

void tab_materials_draw_slots(bool mini) {
	i32 slotw = math_floor(51 * UI_SCALE() + g_config->window_scale * 2);
	i32 num   = math_floor(ui->_window_w / (float)slotw);
	if (num == 0) {
		return;
	}

	bool drag_pos_set = false;
	f32  uix          = 0.0;
	f32  uiy          = 0.0;
	i32  imgw_val     = math_floor(50 * UI_SCALE());

	for (i32 row = 0; row < math_floor(math_ceil(project_materials->length / (float)num)); ++row) {
		i32          mult = g_config->show_asset_names ? 2 : 1;
		f32_array_t *ar   = f32_array_create_from_raw((f32[]){}, 0);
		for (i32 i = 0; i < num * mult; ++i) {
			f32_array_push(ar, 1 / (float)num);
		}
		ui_row(ar);

		ui->_x += 2;
		f32 off = g_config->show_asset_names ? UI_ELEMENT_OFFSET() * 10.0 : 6;
		if (row > 0) {
			ui->_y += off;
		}

		for (i32 j = 0; j < num; ++j) {
			i32 imgw = math_floor(50 * UI_SCALE());
			i32 i    = j + row * num;
			if (i >= project_materials->length) {
				ui_end_element_of_size(imgw);
				if (g_config->show_asset_names) {
					ui_end_element_of_size(0);
				}
				continue;
			}
			gpu_texture_t *img      = UI_SCALE() > 1 ? project_materials->buffer[i]->image : project_materials->buffer[i]->image_icon;
			gpu_texture_t *img_full = project_materials->buffer[i]->image;

			// Highligh selected
			if (g_context->material == project_materials->buffer[i]) {
				if (mini) {
					f32 w = ui->_w / (float)UI_SCALE();
					ui_rect(0, -2, w - 2, w - 4, ui->ops->theme->HIGHLIGHT_COL, 3);
				}
				else {
					i32 off = row % 2 == 1 ? 1 : 0;
					i32 w   = 50 + math_floor(g_config->window_scale * 2);
					ui_fill(-1, -2, w + 3, 2, ui->ops->theme->HIGHLIGHT_COL);
					ui_fill(-1, w - off, w + 3, 2 + off, ui->ops->theme->HIGHLIGHT_COL);
					ui_fill(-1, -2, 2, w + 3, ui->ops->theme->HIGHLIGHT_COL);
					ui_fill(w + 1, -2, 2, w + 4, ui->ops->theme->HIGHLIGHT_COL);
				}
			}

			// Draw material icon
			uix      = ui->_x;
			uiy      = ui->_y;
			i32 tile = UI_SCALE() > 1 ? 100 : 50;
			f32 imgh = mini ? ui_sidebar_default_w_mini * 0.85 * UI_SCALE() : 50 * UI_SCALE();

			if (base_drag_material != NULL && tab_materials_drag_pos == i) {
				ui_fill(-1, -2, 2, imgw_val + 4, ui->ops->theme->HIGHLIGHT_COL);
			}

			ui_state_t state = project_materials->buffer[i]->preview_ready ? ui_image(img, 0xffffffff, imgh)
			                                                               : ui_sub_image(resource_get("icons.k"), 0xffffffff, -1.0, tile, tile, tile, tile);

			if (state == UI_STATE_HOVERED && base_drag_material != NULL) {
				tab_materials_drag_pos = (mouse_x > uix + ui->_window_x + imgw_val / 2.0) ? i + 1 : i;
				drag_pos_set           = true;
			}

			// Draw material numbers when selecting a material via keyboard shortcut
			bool is_typing = ui->is_typing;
			if (!is_typing) {
				if (i < 9 && operator_shortcut(any_map_get(config_keymap, "select_material"), SHORTCUT_TYPE_DOWN)) {
					char *number = i32_to_string(i + 1);
					i32   width  = draw_string_width(ui->ops->font, ui->font_size, number) + 10;
					i32   height = draw_font_height(ui->ops->font, ui->font_size);
					draw_set_color(ui->ops->theme->TEXT_COL);
					draw_filled_rect(uix, uiy, width, height);
					draw_set_color(ui->ops->theme->BUTTON_COL);
					draw_string(number, uix + 5, uiy);
				}
			}

			// Select material
			if (state == UI_STATE_STARTED && ui->input_y > ui->_window_y) {
				if (g_context->material != project_materials->buffer[i]) {
					context_select_material(i);
					if (g_context->tool == TOOL_TYPE_MATERIAL) {
						sys_notify_on_next_frame(&tab_materials_draw_slots_update_fill_layers, NULL);
					}
				}
				base_drag_off_x = -(mouse_x - uix - ui->_window_x - 3);
				base_drag_off_y = -(mouse_y - uiy - ui->_window_y + 1);
				gc_unroot(base_drag_material);
				base_drag_material = g_context->material;
				gc_root(base_drag_material);
				// Double click to show nodes
				if (sys_time() - g_context->select_time < 0.2) {
					ui_base_show_material_nodes();
					gc_unroot(base_drag_material);
					base_drag_material = NULL;
					base_is_dragging   = false;
				}
				g_context->select_time = sys_time();
			}

			// Context menu
			if (ui->is_hovered && ui->input_released_r) {
				context_select_material(i);
				_tab_materials_draw_slots = i;
				ui_menu_draw(&tab_materials_draw_slots_menu, -1, -1);
			}
			if (ui->is_hovered) {
				ui_tooltip_image(img_full, 0);
				if (i < 9) {
					i32 i1 = i + 1;
					ui_tooltip(string("%s - (%s %d)", project_materials->buffer[i]->canvas->name, (char *)any_map_get(config_keymap, "select_material"), i1));
				}
				else {
					ui_tooltip(project_materials->buffer[i]->canvas->name);
				}
			}
			if (g_config->show_asset_names) {
				ui->_x = uix;
				ui->_y += slotw * 0.9;
				ui_text(project_materials->buffer[i]->canvas->name, UI_ALIGN_CENTER, 0x00000000);
				if (ui->is_hovered) {
					if (i < 9) {
						i32 i1 = i + 1;
						ui_tooltip(
						    string("%s - (%s %d)", project_materials->buffer[i]->canvas->name, (char *)any_map_get(config_keymap, "select_material"), i1));
					}
					else {
						ui_tooltip(project_materials->buffer[i]->canvas->name);
					}
				}
				ui->_y -= slotw * 0.9;
				if (i == project_materials->length - 1) {
					ui->_y += j == num - 1 ? imgw : imgw + UI_ELEMENT_H() + UI_ELEMENT_OFFSET();
				}
			}
		}

		ui->_y += mini ? 0 : 6;
	}

	if (base_drag_material != NULL && tab_materials_drag_pos == project_materials->length) {
		ui->_x = uix;
		ui->_y = uiy;
		ui_fill(imgw_val + 1, -2, 2, imgw_val + 4, ui->ops->theme->HIGHLIGHT_COL);
	}

	if (!drag_pos_set) {
		tab_materials_drag_pos = -1;
	}

	bool in_focus = ui->input_x > ui->_window_x && ui->input_x < ui->_window_x + ui->_window_w && ui->input_y > ui->_window_y &&
	                ui->input_y < ui->_window_y + ui->_window_h;
	if (in_focus && ui->is_delete_down && project_materials->length > 1) {
		ui->is_delete_down = false;
		tab_materials_delete_material(g_context->material);
	}
	if (in_focus && ui->is_ctrl_down && ui->is_key_pressed && ui->key_code == KEY_CODE_D) {
		_tab_materials_draw_slots = array_index_of(project_materials, g_context->material);
		sys_notify_on_next_frame(&tab_materials_draw_slots_duplicate, NULL);
	}
	if (in_focus) {
		i32 i = array_index_of(project_materials, g_context->material);
		if (ui->is_key_pressed && ui->key_code == KEY_CODE_UP) {
			if (i > 0) {
				context_select_material(i - 1);
			}
		}
		if (ui->is_key_pressed && ui->key_code == KEY_CODE_DOWN) {
			if (i < project_materials->length - 1) {
				context_select_material(i + 1);
			}
		}
	}
}

void tab_materials_button_new_on_next_frame(void *_) {
	g_context->material = slot_material_create(project_materials->buffer[0]->data, NULL);
	any_array_push(project_materials, g_context->material);
	tab_materials_update_material();
	history_new_material();
}

void tab_materials_button_new(char *text) {
	if (ui_icon_button(text, ICON_PLUS, UI_ALIGN_CENTER)) {
		sys_notify_on_next_frame(&tab_materials_button_new_on_next_frame, NULL);
	}
}

void tab_materials_draw_mini(ui_handle_t *htab) {
	ui_set_hovered_tab_name(tr("Materials"));

	ui_begin_sticky();
	ui_separator(5, true);

	tab_materials_button_nodes();
	tab_materials_button_new("");

	ui_end_sticky();
	ui_separator(3, false);
	tab_materials_draw_slots(true);
}

void tab_materials_draw_full(ui_handle_t *htab) {
	if (ui_tab(htab, tr("Materials"), false, -1, false)) {
		ui_begin_sticky();
		f32_array_t *row = f32_array_create_from_raw(
		    (f32[]){
		        -70,
		        -70,
		        -70,
		    },
		    3);
		ui_row(row);

		tab_materials_button_new(tr("New"));
		if (ui_icon_button(tr("Import"), ICON_IMPORT, UI_ALIGN_CENTER)) {
			project_import_material();
		}
		tab_materials_button_nodes();

		ui_end_sticky();
		ui_separator(3, false);
		tab_materials_draw_slots(false);
	}
}

void tab_materials_draw(ui_handle_t *htab) {
	bool mini = ui->_window_w <= ui_sidebar_w_mini;
	mini ? tab_materials_draw_mini(htab) : tab_materials_draw_full(htab);
}

void tab_materials_accept_material_drop(slot_material_t *material) {
	if (tab_materials_drag_pos == -1) {
		return;
	}

	i32 mat_pos = array_index_of(project_materials, material);
	if (mat_pos != -1 && math_abs(mat_pos - tab_materials_drag_pos) > 0) {
		array_remove(project_materials, material);
		i32 new_pos = tab_materials_drag_pos - mat_pos > 0 ? tab_materials_drag_pos - 1 : tab_materials_drag_pos;
		array_insert(project_materials, new_pos, material);
	}
}

void tab_materials_accept_swatch_drop(swatch_color_t *swatch) {
	g_context->material = slot_material_create(project_materials->buffer[0]->data, NULL);
	for (i32 i = 0; i < g_context->material->canvas->nodes->length; ++i) {
		ui_node_t *node = g_context->material->canvas->nodes->buffer[i];
		if (string_equals(node->type, "RGB")) {
			node->outputs->buffer[0]->default_value = f32_array_create_xyzw(color_get_rb(swatch->base) / 255.0, color_get_gb(swatch->base) / 255.0,
			                                                                color_get_bb(swatch->base) / 255.0, color_get_ab(swatch->base) / 255.0);
		}
		else if (string_equals(node->type, "OUTPUT_MATERIAL_PBR")) {
			node->inputs->buffer[1]->default_value->buffer[0] = swatch->opacity;
			node->inputs->buffer[2]->default_value->buffer[0] = swatch->occlusion;
			node->inputs->buffer[3]->default_value->buffer[0] = swatch->roughness;
			node->inputs->buffer[4]->default_value->buffer[0] = swatch->metallic;
			node->inputs->buffer[7]->default_value->buffer[0] = swatch->height;
		}
	}
	any_array_push(project_materials, g_context->material);
	tab_materials_update_material();
	history_new_material();
}
